File: python/kaczmarzReg.py
import numpy as np
import scipy as sp


def computeRowEnergy(A):
  """
  Calculate the energy of the system matrix rows
  """
  M = A.shape[0]
  energy = np.zeros(M, dtype=np.double)

  for m in range(M):
    energy[m] = np.linalg.norm(A[m,:])
  return energy



def kaczmarzReg(A, b, iterations=10, lambd=0, enforceReal=False, enforcePositive=False, shuffle=False):
  M = A.shape[0]
  N = A.shape[1]

  x = np.zeros(N, dtype=b.dtype)
  residual = np.zeros(M, dtype=x.dtype)

  energy = computeRowEnergy(A)

  rowIndexCycle = np.arange(0,M)

  if shuffle:
    np.random.shuffle(rowIndexCycle)

  lambdIter = lambd

  for l in range(iterations):
    for m in range(M):
      k = rowIndexCycle[m]
      if energy[k] > 0:

        beta = (b[k] - A[k,:].dot(x) - np.sqrt(lambdIter)*residual[k]) / (energy[k]**2 + lambd)

        x[:] +=  beta*A[k,:].conjugate()

        residual[k] += np.sqrt(lambdIter) * beta

    if enforceReal and np.iscomplexobj(x):
      x.imag = 0
    if enforcePositive:
      x = x * (x.real > 0)

  return x

File: julia/kaczmarzReg.jl
"""
The regularized kaczmarz algorithm solves the Thikonov regularized least squares Problem
argminₓ(‖Ax-b‖² + λ‖b‖²).

# Arguments

* `A::AbstractMatrix`: System matrix A
* `b::Vector`: Measurement vector b
* `iterations::Int`: Number of iterations of the iterative solver
* `lambd::Float64`: The regularization parameter, relative to the matrix trace
* `shuff::Bool`: Enables random shuffeling of rows during iterations in the kaczmarz algorithm
* `enforceReal::Bool`: Enable projection of solution on real plane during iteration
* `enforcePositive::Bool`: Enable projection of solution onto positive halfplane during iteration
"""
function kaczmarzReg(A::AbstractMatrix{T}, b::Vector{T}, iterations, lambd, shuff,
                     enforceReal, enforcePositive) where T
  M = size(A,2)
  N = size(A,1)

  x = zeros(T, N)
  residual = zeros(T, M)

  energy = zeros(Float64, M)
  for m=1:M
    energy[m] = norm(A[:,m])
  end

  rowIndexCycle = collect(1:M)

  if shuff
    shuffle(rowIndexCycle)
  end

  lambdIter = lambd

  for l=1:iterations
    for m=1:M
      k = rowIndexCycle[m]
      if energy[k] > 0
        tmp = dot_with_matrix_row(A,x,k)

        beta = (b[k] - tmp - sqrt(lambdIter)*residual[k]) / (energy[k]^2 + lambd)

        for n=1:size(A,1)
          x[n] += beta*conj(A[n,k])
        end

        residual[k] = residual[k] + beta*sqrt(lambdIter)
      end
    end

    if enforceReal && eltype(x) <: Complex
      x = complex.(real.(x), 0)
    end
    if enforcePositive
      x[real.(x) .< 0] .= 0
    end
  end

  return x
end

"""
Calculates the dot product between x and the k-th matrix row of A.

# Arguments

* `A::AbstractMatrix`: System matrix A
* `x::Vector`: Measurement vector b
* `k::Int: matrix row
"""
function dot_with_matrix_row(A, x, k)
  tmp = 0.0
  for n=1:size(A,1)
    tmp += A[n,k]*x[n]
  end
  tmp
end
